scripts/cc-cli/src/test/integration-tests/collectCoins.test.ts

```diff
@@ -0,0 +1,142 @@
+  );
+
+  beforeAll(async () => {
+    await deployCtcToken((global as any).CREDITCOIN_CTC_CONTRACT_ADDRESS);
+    setArg(
+      "CREDITCOIN_CTC_CONTRACT_ADDRESS",
+      process.env.CREDITCOIN_CTC_CONTRACT_ADDRESS,
+    );
+    setArgIfUndef(
+      "CREDITCOIN_CTC_BURN_TX_HASH",
+      process.env.CREDITCOIN_CTC_BURN_TX_HASH,
+    );
+
+    await retry(5, async (resolve: any, reject: any) => {
+      await setAuthorities().then(resolve).catch(reject);
+    }).catch(() => {
+      console.log("Could not setup testing authorities");
+      process.exit(1);
+    });
+
+    await cryptoWaitReady();
+
+    caller = randomTestAccount(false);
+
+    ccApi = await creditcoinApi((global as any).CREDITCOIN_API_URL);
+    sudo = arg("CREDITCOIN_CREATE_SIGNER")(keyring, "sudo");
+
+    const { api } = ccApi;
+
+    /* eslint-disable @typescript-eslint/naming-convention */
+    const contract = api.createType(
+      "PalletCreditcoinOcwTasksCollectCoinsDeployedContract",
+      {
+        address: (global as any).CREDITCOIN_CTC_CONTRACT_ADDRESS,
+        chain: blockchain,
+      },
+    );
+
+    await api.tx.sudo
+      .sudo(api.tx.creditcoin.setCollectCoinsContract(contract))
+      .signAndSend(sudo, { nonce: -1 });
+  }, 100_000);
+
+  afterAll(async () => {
+    await ccApi.api.disconnect();
+  });
+
+  test("e2e", async () => {
+    const {
+      api,
+      utils: { signAccountId },
+    } = ccApi;
+
+    const provider = new providers.JsonRpcProvider(
+      arg("CREDITCOIN_ETHEREUM_NODE_URL"),
+    );
+    const deployerWallet = new Wallet(
+      arg("CREDITCOIN_CTC_DEPLOYER_PRIVATE_KEY"),
+      provider,
+    );
+
+    const fundTx = await fundFromSudo(
+      caller.address,
+      parseAmountInternal("5"),
+      arg("CREDITCOIN_API_URL"),
+    );
+    await signSendAndWatch(fundTx, api, sudo);
+
+    await tryRegisterAddress(
+      ccApi,
+      deployerWallet.address,
+      blockchain,
+      signAccountId(deployerWallet, caller.address),
+      caller.keyring,
+      (global as any).CREDITCOIN_REUSE_EXISTING_ADDRESSES,
+    );
+
+    // Read balance after register address call but prior to collect coins
+    const starting = await getBalance(caller.address, api);
+
+    const collectResult = execa.commandSync(
+      `npx creditcoin-cli collect-coins -u ${
+        arg("CREDITCOIN_API_URL") as string
+      }`,
+      {
+        env: {
+          EXTERNAL_ADDR: deployerWallet.address,
+          BURN_TX_HASH: arg("CREDITCOIN_CTC_BURN_TX_HASH"),
+          CC_SECRET: caller.secret,
+        },
+      },
+    );
+
+    const collectOutput = collectResult.stdout.split("\n");
+    expect(collectResult.failed).toBe(false);
+    expect(collectResult.exitCode).toBe(0);
+    expect(collectResult.stderr).toBe("");
+    expect(collectOutput[collectOutput.length - 1]).toBe("Success!");
+
+    // note: wait before fetching the balance again. Otherwise we get bogus values
+    await new Promise((f) => setTimeout(f, 3000));
```

Comment:
I thought about it some more and this still feels weird to me. I don't have any answers, but really as soon as the collect coins has been verified the balance should reflect the minted coins. So unless polkadot JS is quietly using finalized state or something, or there's a bug in polkadot JS, it doesn't really make sense to me that you would be getting stale values.

Anything involving polkadot JS internals is a bit of a (massive) rabbit hole though, so I'm okay with leaving this a mystery for now. But FTR it still doesn't sit right with me that we have to wait like this at all

Reply:
FTR this is already converted to waiting for a number of blocks (2), which is better than an arbitrary amount in seconds: https://github.com/gluwa/creditcoin/pull/1325/commits/b62fa811594cbf4a2bca6309cf201604bd055053